### PR TITLE
Implements TemplateClass.

### PR DESCRIPTION
**Pull request by tomsons26**

### CODE REVIEW

File: src/game/engine/templatetype.h [hunk 1]
```diff
@@ -462,6 +462,8 @@ class TemplateTypeClass : public ObjectTypeClass
     int Get_Width() { return Width; }
     int Get_Height() { return Height; }
 
+    TemplateType Get_Type() { return Type; }
```
Comment (ghost) [inline src/game/engine/templatetype.h]: This function needs to be "const" as we are not modifying its contents.

File: src/game/engine/template.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+#ifndef TEMPLATE_H
+#define TEMPLATE_H
+
+#include "always.h"
+#include "object.h"
+#include "templatetype.h"
+
+#ifndef GAME_DLL
+TFixedIHeapClass<TemplateClass> g_Templates;
```
Comment (ghost) [inline src/game/engine/template.h]: This needs to be moved to the bottom of the file as TemplateClass is not yet defined.

File: src/game/engine/template.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    TemplateClass(const TemplateClass &that);
+    TemplateClass(const NoInitClass &noinit) : ObjectClass(noinit) {}
+    virtual ~TemplateClass(void);
+
+    void *operator new(size_t size);
+    void *operator new(size_t size, void *ptr) { return ptr; }
+    void operator delete(void *ptr);
+#ifndef COMPILER_WATCOM // Watcom doesn't like this, MSVC/GCC does.
+    void operator delete(void *ptr, void *place) {}
+#endif
```
Comment (ghost) [inline src/game/engine/template.h:37]: Not an issue, but a style thing. Space after this block of definitions.

File: src/game/engine/template.h [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    virtual TemplateTypeClass &Class_Of() const override { return *m_Type; }
+    virtual void Draw_It(int x, int y, WindowNumberType window) const override{};
+    virtual BOOL Mark(MarkType mark) override;
+
+    TemplateType Get_Type() { return m_Type->Get_Type(); }
```
Comment (ghost) [inline src/game/engine/template.h]: This function needs to be "const" as we are not modifying its contents.

File: src/game/engine/template.cpp [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+ *
+ * 0x00569F24
+ */
+BOOL TemplateClass::Mark(MarkType mark)
+{
+    DEBUG_ASSERT(m_IsActive);
+    static bool _noup = false;
+
+    void *image = Class_Of().Get_Image_Data();
+    if (image == nullptr || ObjectClass::Mark(mark) == false) {
```
Comment (ghost) [inline src/game/engine/template.cpp]: Your formatting of the check against ObjectClass::Mark() here is overkill. Use "!ObjectClass::Mark()".

File: src/game/engine/template.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+TFixedIHeapClass<TemplateClass> g_Templates;
+#endif
+
+class TemplateClass : public ObjectClass
+{
+public:
+    TemplateClass(TemplateType type, cell_t cellnum);
+    TemplateClass(const TemplateClass &that);
+    TemplateClass(const NoInitClass &noinit) : ObjectClass(noinit) {}
+    virtual ~TemplateClass(void);
```
Comment (ghost) [inline src/game/engine/template.h]: Destructor here has C style "(void)", we don't need this in C++.

File: src/game/engine/template.cpp [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+            }
+        }
+    }
+    if (mark == MARK_PUT) {
+        _noup = true;
+        delete this;
+        _noup = false;
+    }
+    return true;
+}
```
Comment (ghost) [inline src/game/engine/template.cpp]: Always make sure files have at least one blank new line at the end.

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]

File: src/game/engine/template.cpp [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+/**
+ * @brief
+ *
+ * 0x0056A198
+ */
+void *TemplateClass::operator new(size_t size)
+{
+    TemplateClass *this_ptr = g_Templates.Alloc();
+    DEBUG_ASSERT(this_ptr != nullptr);
```
Comment (OmniBlade) [inline src/game/engine/template.cpp:46]: You shouldn't really assert something that your function actually expects to happen and has a handler for. Asserts are for catching stuff that your function is written to never expect to happen.

**Review by OmniBlade — CHANGES_REQUESTED** [review]

**Review by OmniBlade — APPROVED** [review]

**Pull request merged**